### PR TITLE
temporal alignment for raster stacker

### PR DESCRIPTION
**Pull request by michaelmattig**
- [ ] I added an entry to [`CHANGELOG.md`](CHANGELOG.md) if knowledge of this change could be valuable to users.

---

Here is a brief summary of what I did:

<TEXT>

### CODE REVIEW

**Comment by michaelmattig** [comment]
[Multibands: RasterStacker soll temporal alignen](https://trello.com/c/g7cZmMxo/1505-multibands-rasterstacker-soll-temporal-alignen)

File: operators/src/adapters/simple_raster_stacker.rs [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// Stacks the bands of the input raster streams to create a single raster stream with all the combined bands.
+///
+/// NOTE: ALL THE INPUT STREAMS MUST HAVE THE SAME TIME INTERVALS.
+///
+/// If the streams are not guaranteed to be temporally aligned, use the `RasterStackerAdapter` instead.
+#[pin_project(project = SimpleRasterStackerAdapterProjection)]
+pub struct SimpleRasterStackerAdapter<S> {
```
Comment (jdroenner) [inline operators/src/adapters/simple_raster_stacker.rs:18]: Maybe call it "AlignedRasterStackerAdapter"?

**Review by jdroenner — COMMENTED** [review]
I think we need more tests:
- Source with 1 BAnd and Source with more then 2
- 3 Sources

File: operators/src/adapters/raster_time.rs [hunk 2]
```diff
@@ -177,6 +178,7 @@ where
     F::Output: Future<Output = Result<F::Stream>>,
 {
     pub fn new(sources: [F; N], query_rect: RasterQueryRectangle) -> Self {
+        // TODO: need to ensure all sources are single-band
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/adapters/raster_time.rs:181]: Yes, the processors would need additional methods, like I proposed in the follow up task.

Also I am not sure if we even still need this adapter. Once we update the Expression and the Rgb stuff to use the Stacker, it is no longer used anywhere else.

File: operators/src/adapters/raster_stacker.rs [hunk 3]
```diff
@@ -1,140 +1,358 @@
...
+                    StreamState::ProducingTimeSlice {
+                        first_tiles,
+                        time_slice,
+                        current_stream,
+                        current_band,
+                        current_spatial_tile,
+                    } => {
+                        let tile = if *current_spatial_tile == 0 && *current_band == 0 {
+                            // consume tiles that were already computed first
+                            Some(Ok(first_tiles[*current_stream].clone())) // TODO: avoid clone and instead consume the tile
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/adapters/raster_stacker.rs:262]: either that or just remove it from the Vec and keep track of the indexes.

File: operators/src/adapters/raster_stacker.rs [hunk 4]
```diff
@@ -1,140 +1,358 @@
...
+                        } else {
+                            ready!(Pin::new(&mut streams[*current_stream]).poll_next(cx))
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/adapters/raster_stacker.rs:264]: I don't think we can do this. The adapter consumes band for band, tile for tile. So if there is more than one band for a source, we poll it multiple times before going to the next source.

File: operators/src/adapters/simple_raster_stacker.rs (hunk 1 above)
Comment (michaelmattig) [reply operators/src/adapters/simple_raster_stacker.rs:18]: Is that really better? One might mistake it for the one aligning stuff.

File: operators/src/adapters/raster_stacker.rs [hunk 5]
```diff
@@ -1,140 +1,358 @@
...
-                                found: tile.time,
-                            },
-                        )));
+        loop {
+            match state.as_mut().project() {
+                ArrayStateProjection::Initial => {
+                    let array_of_futures = sources
+                        .iter()
+                        .map(|source| {
+                            let mut query_rect = query_rect.clone();
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/adapters/raster_stacker.rs]: cab0f1883e9802cb33206622bad2542a8e1362da

**Comment by michaelmattig** [comment]
> I think we need more tests:
> 
>     * Source with 1 BAnd and Source with more then 2
> 
>     * 3 Sources

073339d32d54bb99299fc804276379161b542671

**Review by jdroenner — APPROVED** [review]

**Pull request merged**